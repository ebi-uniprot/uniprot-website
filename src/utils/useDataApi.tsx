import React, { useState, useEffect } from 'react';
import fetchData from './fetchData';
import 'regenerator-runtime/runtime';

const useDataApi = (url: string) => {
  const [data, setData] = useState<any>({});

  const getData = async () => {
    const result = await fetchData(url);
    setData(result.data);
  };

  useEffect(() => {
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
