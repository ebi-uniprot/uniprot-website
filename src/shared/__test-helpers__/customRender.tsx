/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
