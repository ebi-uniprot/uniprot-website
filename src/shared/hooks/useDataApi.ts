import { useEffect, useReducer, useRef } from 'react';
import axios, {
  AxiosResponse,
  AxiosError,
  AxiosRequestConfig,
  AxiosProgressEvent,
  isCancel,
} from 'axios';
import joinUrl from 'url-join';

import useMessagesDispatch from './useMessagesDispatch';

import fetchData from '../utils/fetchData';
import { addMessage } from '../../messages/state/messagesActions';
import * as logging from '../utils/logging';
import { apiPrefix } from '../config/apiUrls/apiPrefix';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Namespace } from '../types/namespaces';
import { UserPreferenceKey } from './useLocalStorage';

export type CustomError = AxiosError<undefined | { messages?: string[] }>;

const invalidFieldMessage = /Invalid fields parameter value '(?<field>[^']*)'/;
const namespacedURL = new RegExp(
  joinUrl(
    apiPrefix,
    `(?<namespace>${Object.values(Namespace).join('|')})`,
    'search'
  )
);

// TODO: possible improvement:
// See https://developers.google.com/web/tools/workbox/modules/workbox-broadcast-update#using_broadcast_update
// Listen for a service worker message saying that there is new data to get
// fresher data from cache and replace the data within this custom hook

// Threshold under which a progress event will not be emitted
// This is in order to avoid unecessary renders
const PROGRESS_THRESHOLD = 2_000; // 2s

export type UseDataAPIState<T> = {
  loading: boolean;
  progress?: number;
  data?: T;
  status?: AxiosResponse['status'];
  statusText?: AxiosResponse['statusText'];
  headers?: AxiosResponse['headers'];
  error?: CustomError;
  redirectedTo?: string;
  url?: string | null;
};

enum ActionType {
  INIT = 'INIT',
  PROGRESS = 'PROGRESS',
  SUCCESS = 'SUCCESS',
  ERROR = 'ERROR',
}

type Action<T> =
  | { type: ActionType.INIT; url?: string }
  | { type: ActionType.PROGRESS; progress: number }
  | {
      type: ActionType.SUCCESS;
      response?: AxiosResponse<T>;
      originalURL?: string;
      progress?: 1;
    }
  | { type: ActionType.ERROR; error: CustomError };

const createReducer =
  <T>() =>
  (state: UseDataAPIState<T>, action: Action<T>): UseDataAPIState<T> => {
    // eslint-disable-next-line default-case
    switch (action.type) {
      case ActionType.INIT:
        return {
          loading: true,
          url: action.url,
        };
      case ActionType.PROGRESS:
        return {
          ...state,
          progress: action.progress,
        };
      case ActionType.SUCCESS:
        // eslint-disable-next-line no-case-declarations
        const newState: UseDataAPIState<T> = {
          ...state,
          loading: false,
          progress: action.progress,
          data: action.response?.data,
          status: action.response?.status,
          statusText: action.response?.statusText,
          headers: action.response?.headers,
        };
        if (
          action.response?.request?.responseURL &&
          action.response.request.responseURL !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseURL;
        } else if (
          // Issue with casing in axios-mock-adapter?
          action.response?.request?.responseUrl &&
          action.response.request.responseUrl !== action.originalURL
        ) {
          newState.redirectedTo = action.response.request.responseUrl;
        }
        return newState;
      case ActionType.ERROR:
        return {
          ...state,
          loading: false,
          progress: undefined,
          status: action.error.response?.status,
          statusText: action.error.response?.statusText,
          headers: action.error.response?.headers,
          error: action.error,
        };
    }
  };

function useDataApi<T>(
  url?: string | null,
  options?: AxiosRequestConfig
): UseDataAPIState<T> {
  const [state, dispatch] = useReducer(createReducer<T>(), { loading: !!url });
  const messagesDispatch = useMessagesDispatch();
  const optionsRef = useRef(options);

  useEffect(() => {
    if (options !== optionsRef.current) {
      optionsRef.current = options;
    }
  }, [options]);

  useEffect(() => {
    // need this variable to ensure state updates don't occur when cancelled/unmounted
    // https://github.com/facebook/react/issues/14369#issuecomment-468267798
    let didCancel = false;

    // we don't require a URL, we just don't need data anymore
    // assume succes with no data
    if (!url) {
      dispatch({ type: ActionType.SUCCESS });
      return;
    }

    dispatch({ type: ActionType.INIT, url });

    // variables to handle cancellation
    // eslint-disable-next-line import/no-named-as-default-member
    const source = axios.CancelToken.source();

    // to keep track of the last time we dispatched a progress action
    let lastProgressDate: number;
    // actual request
    fetchData<T>(url, source.token, {
      ...optionsRef.current,
      onDownloadProgress: (progressEvent: AxiosProgressEvent) => {
        optionsRef.current?.onDownloadProgress?.(progressEvent);
        if (didCancel || progressEvent.progress === undefined) {
          return;
        }
        const now = Date.now();
        // If last time we dispatched a progress action was less than threshold
        if (lastProgressDate && now - lastProgressDate < PROGRESS_THRESHOLD) {
          // skip, don't want to refresh too often
          return;
        }
        dispatch({
          type: ActionType.PROGRESS,
          progress: progressEvent.progress,
        });
        lastProgressDate = now;
      },
    }).then(
      // handle ok
      (response: AxiosResponse<T>) => {
        if (didCancel) {
          return;
        }
        dispatch({
          type: ActionType.SUCCESS,
          response,
          originalURL: url,
          progress: 1,
        });
      },
      // catch error
      (error: CustomError) => {
        // Ignore 404 because it might mean "valid request but no data"
        if (error.response?.status && error.response.status !== 404) {
          logging.error(error, { tags: { origin: 'useDataApi', url } });
        }
        if (isCancel(error) || didCancel) {
          return;
        }
        dispatch({ type: ActionType.ERROR, error });
      }
    );

    // handle unmounting of the hook

    return () => {
      source.cancel();
      didCancel = true;
    };
  }, [url, optionsRef]);

  useEffect(() => {
    // TODO: when refactoring, accept a onError callback to do this from outside
    if (state.status === 400) {
      // Just in case a field is updated in the server and is no longer valid
      try {
        for (const message of state.error?.response?.data?.messages || []) {
          const messageMatch = message.match(invalidFieldMessage);
          const invalidField = messageMatch?.groups?.field;
          if (!messageMatch) {
            continue;
          }
          const nsMatch = state.url?.match(namespacedURL);
          if (!nsMatch?.groups?.namespace) {
            continue;
          }
          const key =
            `table columns for ${nsMatch?.groups?.namespace}` as UserPreferenceKey;
          const faultyArray: string[] = JSON.parse(
            localStorage.getItem(key) as string
          );
          const noFaults = faultyArray.filter(
            (column) =>
              // Clean up any wrong type of value, and the faulty value too
              column && typeof column === 'string' && column !== invalidField
          );
          // If no valid columns then store undefined which will cause the
          // default columns to be used.
          const correctArray = JSON.stringify(
            noFaults.length ? noFaults : undefined
          );
          localStorage.setItem(key, correctArray);
          // Signals to all useLocalStorage hooks in use to rerender
          window.dispatchEvent(
            new StorageEvent('storage', {
              key,
              newValue: correctArray,
              storageArea: window.localStorage,
            })
          );
          return;
        }
      } catch {
        /**/
      }
      const messageContent =
        state.error?.response?.data?.messages?.join(',') || '400 Error';
      messagesDispatch(
        addMessage({
          id: messageContent,
          content: messageContent,
          format: MessageFormat.POP_UP,
          level: MessageLevel.FAILURE,
          displayTime: 5_000,
        })
      );
    }
  }, [messagesDispatch, state.status, state.error, state.url]);

  // when changing the URL, the state is set asynchronously, this is to set it
  // to loading synchronously to avoid using previous data
  if (state.url !== url) {
    return { loading: Boolean(url), url };
  }

  return state;
}

export default useDataApi;
