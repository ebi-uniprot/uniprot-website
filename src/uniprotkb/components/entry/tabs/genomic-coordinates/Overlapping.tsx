import { Fragment } from 'react';
import { ExpandableList, SpinnerIcon } from 'franklin-sites';
import { Link } from 'react-router-dom';

import ExternalLink from '../../../../../shared/components/ExternalLink';
import GenomicLoc, { getEnsemblLink } from './GenomicLoc';

import useDataApi from '../../../../../shared/hooks/useDataApi';

import { getEntryPath } from '../../../../../app/config/urls';
import { proteinsApi } from '../../../../../shared/config/apiUrls';

import { Namespace } from '../../../../../shared/types/namespaces';
import { TabLocation } from '../../Entry';
import { GenomicEntry } from './types';

type OverlappingProps = {
  taxID: number;
  chromosome?: string;
  start: number;
  end: number;
  currentEntry: string;
};

const Overlapping = ({
  taxID,
  chromosome,
  start,
  end,
  currentEntry,
}: OverlappingProps) => {
  const sp = new URLSearchParams({
    taxid: `${taxID}`,
    location: `${start}-${end}`,
  });
  if (chromosome) {
    sp.set('chromosome', chromosome);
  }

  /** Fetch all entries for which part of the genome location overlaps (even
   * partially) to the current one */
  const { loading, data, error, status } = useDataApi<Array<GenomicEntry>>(
    `${proteinsApi.coordinates()}?${sp}`
  );

  if (loading) {
    return <SpinnerIcon width="1em" height="1em" />;
  }

  if (error && status !== 404) {
    return <em>An error happened</em>;
  }

  // Removes the original protein from the list of proteins at these coordinates
  const otherEntries = data?.filter(
    (entry) => entry.accession !== currentEntry
  );

  if (!otherEntries?.length) {
    return <>none</>;
  }

  // Gather minmax coordinates of overlapping entries
  let min = start;
  let max = end;
  const entryList: JSX.Element[] = [];

  // Render an entry link for each overlapping entry
  for (const entry of otherEntries) {
    const entryAndLocations = [
      <Link
        to={getEntryPath(
          Namespace.uniprotkb,
          entry.accession,
          TabLocation.Entry
        )}
        key={entry.accession}
      >
        {entry.accession}
      </Link>,
      ' (',
    ];
    // Render an Ensembl link for each overlapping entry's overlapping location
    for (const { genomicLocation } of entry.gnCoordinate || []) {
      if (genomicLocation.chromosome !== chromosome) {
        // This location not even within the original chromosome, skip it
        continue; // eslint-disable-line no-continue
      }
      const otherStart = genomicLocation.reverseStrand
        ? genomicLocation.end
        : genomicLocation.start;
      const otherEnd = genomicLocation.reverseStrand
        ? genomicLocation.start
        : genomicLocation.end;
      if (otherStart === undefined || otherEnd === undefined) {
        continue; // eslint-disable-line no-continue
      }
      const startWithinRange = otherStart >= start && otherStart <= end;
      const endWithinRange = otherEnd >= start && otherEnd <= end;
      const startBeforeRange = otherStart <= start;
      const endAfterRange = otherEnd >= end;
      if (
        // No part within the range
        !(startWithinRange || endWithinRange) &&
        // This range does not include the current entry range
        !(startBeforeRange && endAfterRange)
      ) {
        // This location not within the original location range, skip it
        continue; // eslint-disable-line no-continue
      }
      if (entryAndLocations.length > 2) {
        entryAndLocations.push(', ');
      }
      entryAndLocations.push(
        <GenomicLoc key="gl" genomicLocation={genomicLocation} taxID={taxID} />
      );
      if (otherStart < min) {
        min = otherStart;
      }
      if (otherEnd > end) {
        max = otherEnd;
      }
    }
    entryAndLocations.push(')');
    entryList.push(
      <Fragment key={entry.accession}>{entryAndLocations}</Fragment>
    );
  }

  return (
    <>
      <ExpandableList descriptionString="entries" displayNumberOfHiddenItems>
        {entryList}
      </ExpandableList>
      {/* If the overlapping entries expand further than the current one, render
      a link to the genome browser to explore the full range */}
      {(min !== start || max !== end) && (
        <ExternalLink url={getEnsemblLink(taxID, chromosome, min, max)}>
          Explore genomic region in Ensembl
        </ExternalLink>
      )}
    </>
  );
};

export default Overlapping;
