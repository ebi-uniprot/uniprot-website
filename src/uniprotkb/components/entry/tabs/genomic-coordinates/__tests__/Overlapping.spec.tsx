/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { render } from '@testing-library/react';

import customRender from '../../../../../../shared/__test-helpers__/customRender';

import Overlapping from '../Overlapping';

import useDataApi from '../../../../../../shared/hooks/useDataApi';

import O00560 from './__mocks__/O00560';

jest.mock('../../../../../../shared/hooks/useDataApi');

describe('VariationViewer component', () => {
  it('renders on loading', () => {
    (useDataApi as jest.Mock).mockReturnValue({ loading: true });
    const { asFragment } = render(
      <Overlapping
        taxID={O00560.taxid!}
        chromosome={O00560.gnCoordinate![0].genomicLocation.chromosome}
        start={O00560.gnCoordinate![0].genomicLocation.start!}
        end={O00560.gnCoordinate![0].genomicLocation.end!}
        currentEntry="O00560"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders on error', () => {
    (useDataApi as jest.Mock).mockReturnValue({
      loading: false,
      error: new Error('some error'),
      status: 500,
    });
    const { asFragment } = render(
      <Overlapping
        taxID={O00560.taxid!}
        chromosome={O00560.gnCoordinate![0].genomicLocation.chromosome}
        start={O00560.gnCoordinate![0].genomicLocation.start!}
        end={O00560.gnCoordinate![0].genomicLocation.end!}
        currentEntry="O00560"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders on no data', () => {
    (useDataApi as jest.Mock).mockReturnValue({
      loading: false,
      status: 404,
    });
    const { asFragment } = render(
      <Overlapping
        taxID={O00560.taxid!}
        chromosome={O00560.gnCoordinate![0].genomicLocation.chromosome}
        start={O00560.gnCoordinate![0].genomicLocation.start!}
        end={O00560.gnCoordinate![0].genomicLocation.end!}
        currentEntry="O00560"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders on data, no other entry at the same position', () => {
    (useDataApi as jest.Mock).mockReturnValue({
      loading: false,
      data: [O00560],
      status: 200,
    });
    const { asFragment } = customRender(
      <Overlapping
        taxID={O00560.taxid!}
        chromosome={O00560.gnCoordinate![0].genomicLocation.chromosome}
        start={O00560.gnCoordinate![0].genomicLocation.start!}
        end={O00560.gnCoordinate![0].genomicLocation.end!}
        currentEntry="O00560"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders on data, other than current entry', () => {
    (useDataApi as jest.Mock).mockReturnValue({
      loading: false,
      data: [O00560],
      status: 200,
    });
    const { asFragment } = customRender(
      <Overlapping
        taxID={O00560.taxid!}
        chromosome={O00560.gnCoordinate![0].genomicLocation.chromosome}
        start={O00560.gnCoordinate![0].genomicLocation.start!}
        end={O00560.gnCoordinate![0].genomicLocation.end!}
        currentEntry="PXXXXX"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
