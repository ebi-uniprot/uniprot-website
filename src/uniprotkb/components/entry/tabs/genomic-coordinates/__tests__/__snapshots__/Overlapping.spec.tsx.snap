// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariationViewer component renders on data, no other entry at the same position 1`] = `
<DocumentFragment>
  none
</DocumentFragment>
`;

exports[`VariationViewer component renders on data, other than current entry 1`] = `
<DocumentFragment>
  <ul
    class="expandable-list no-bullet"
  >
    <li>
      <a
        href="/uniprotkb/O00560/entry"
      >
        O00560
      </a>
       (
      <a
        class="external-link"
        href="https://www.ensembl.org/9606/Location/View?r=8:58565034-58581737"
        rel="noopener"
        target="_blank"
      >
        8:58,565,034-58,581,737
        <test-file-stub
          data-testid="external-link-icon"
          width="12.5"
        />
      </a>
      , 
      <a
        class="external-link"
        href="https://www.ensembl.org/9606/Location/View?r=8:58565034-58581737"
        rel="noopener"
        target="_blank"
      >
        8:58,565,034-58,581,737
        <test-file-stub
          data-testid="external-link-icon"
          width="12.5"
        />
      </a>
      )
    </li>
  </ul>
</DocumentFragment>
`;

exports[`VariationViewer component renders on error 1`] = `
<DocumentFragment>
  <em>
    An error happened
  </em>
</DocumentFragment>
`;

exports[`VariationViewer component renders on loading 1`] = `
<DocumentFragment>
  <test-file-stub
    height="1em"
    width="1em"
  />
</DocumentFragment>
`;

exports[`VariationViewer component renders on no data 1`] = `
<DocumentFragment>
  none
</DocumentFragment>
`;
